File: python/2.2.7/2.2.7_RFID.py
#!/usr/bin/env python3

import RPi.GPIO as GPIO
import MFRC522
import sys
import os

# Create an object of the class MFRC522
mfrc = MFRC522.MFRC522()

def setup():
	print ("Program is starting ... "	)
	print ("Press Ctrl-C to exit.\n")
	pass
	
def loop():
	global mfrc
	while(True):
		print ("Scanning Card... ")
		mfrc = MFRC522.MFRC522()
		isScan = True
		while isScan:
			# Scan for cards   
			(status,TagType) = mfrc.MFRC522_Request(mfrc.PICC_REQIDL)
			# Get the UID of the card
			(status,uid) = mfrc.MFRC522_Anticoll()
			# If we have the UID, continue
			if status == mfrc.MI_OK:
				# Select the scanned tag
				if mfrc.MFRC522_SelectTag(uid) == 0:
					print ("MFRC522_SelectTag Failed!")
				if cmdloop(uid) < 1 :
					isScan = False

				
def cmdloop(cardID):
	blockAddr = 2
	while(True):
		print ("Reading card %2X%2X%2X%2X%2X > \n"%(cardID[0],cardID[1],cardID[2],cardID[3],cardID[4]))
		key = [0xFF,0xFF,0xFF,0xFF,0xFF,0xFF]
		# Authenticate
		status = mfrc.MFRC522_Auth(mfrc.PICC_AUTHENT1A, blockAddr, key, cardID)
		# Check if authenticated
		if status == mfrc.MI_OK:
			mfrc.MFRC522_Read(blockAddr)
			print(" ")
		else:
			print ("Authentication error")
			return 0
		print ("\nInput'write' or 'clean' to modify the data\n>",end="")
		inCmd = input()
		cmd = inCmd.split(" ")

		if cmd[0] == "write":
			print ("\nInput the Data (less than 16 character)\n>",end="")
			inCmd = input()
			cmd = inCmd.split(" ")			
			data = [0]*16
			data = cmd[0][0:17]
			data = map(ord,data)
			data = list(data)
			lenData = len(list(data))
			if lenData<16:
				data+=[0]*(16-lenData)
			# This is the default key for authentication
			key = [0xFF,0xFF,0xFF,0xFF,0xFF,0xFF]			
			# Authenticate
			status = mfrc.MFRC522_Auth(mfrc.PICC_AUTHENT1A, blockAddr, key, cardID)
			# Check if authenticated
			if status == mfrc.MI_OK:
				mfrc.MFRC522_Write(blockAddr, data)
			else:
				print ("Authentication error")
				return 0
			
		elif cmd[0] == "clean":
			data = [0]*16
			# This is the default key for authentication
			key = [0xFF,0xFF,0xFF,0xFF,0xFF,0xFF]			
			# Authenticate
			status = mfrc.MFRC522_Auth(mfrc.PICC_AUTHENT1A, blockAddr, key, cardID)
			# Check if authenticated
			if status == mfrc.MI_OK:
				mfrc.MFRC522_Write(blockAddr, data)
			else:
				print ("Authentication error")
				return 0
		else :
			return 0
				
def destroy():
	GPIO.cleanup()

if __name__ == "__main__":
	setup()
	try:
		loop()
	except KeyboardInterrupt:  # Ctrl+C captured, exit
		destroy()

File: c/2.2.7/2.2.7_RFID.c
/**********************************************************************
* Filename    : 2.2.7_RFID.c
* Description : Use MFRC522 Read and Write Mifare Card.
* Author      : Robot
* E-mail      : support@sunfounder.com
* website     : www.sunfounder.com
* Update      : Daisy    2019/08/01
**********************************************************************/

#include <stdio.h>
#include <stdint.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <stdlib.h>
#include "mfrc522.h"

int scan_loop(uint8_t *CardID);
int tag_select(uint8_t *CardID);

int main(int argc, char **argv) {
	MFRC522_Status_t ret;
	uint8_t CardID[5] = { 0x00, };
	uint8_t tagType[16] = {0x00,};
	static char command_buffer[1024];

	ret = MFRC522_Init('B');
	if (ret < 0) {
		perror("Failed to initialize");
		exit(-1);
	}

	while (1) {
		puts("Scanning Card... ");
		while (1) {
			ret = MFRC522_Request(PICC_REQIDL, tagType);
			if (ret == MI_OK) {
				ret = MFRC522_Anticoll(CardID);	
				if(ret == MI_OK){
					ret = tag_select(CardID);
					if (ret == MI_OK) {
						ret = scan_loop(CardID);
						if (ret < 0) {
							printf("Error...\r\n");
							break;
						} 
					}
				}
				else{
					printf("Get Card ID failed!\r\n");
				}								
			}
			MFRC522_Halt();
		}
		MFRC522_Halt();
		MFRC522_Init('B');
	}
}
int scan_loop(uint8_t *CardID) {
	while (1) {
		char input[32];
		int block_start=2;
		printf("Reading card %02X%02X%02X%02X \n", CardID[0], CardID[1], CardID[2], CardID[3]);
		MFRC522_Debug_DumpSector(CardID, block_start);
		printf("Input 'write' or 'clean' to modify the data\n >");
		scanf("%s", input);
		printf("\n");
		if(strcmp(input, "clean") == 0){
			getchar();
			if (MFRC522_Debug_Clean(CardID, block_start)) {
				return -1;
			}
			
		} else if (strcmp(input, "write") == 0) {
			char write_buffer[256];
			size_t len = 0;
			getchar();
			printf("Input the Data (less than 16 character)\n>");
			scanf("%s",write_buffer);
			if (MFRC522_Debug_Write(CardID, block_start, write_buffer,
					strlen(write_buffer)) < 0) {
				return -1;
			}
		} 
	}
	return 0;
}
int tag_select(uint8_t *CardID) {
	int ret_int;
	ret_int = MFRC522_SelectTag(CardID);
	if (ret_int == 0) {
		printf("Card Select Failed\r\n");
		return -1;
	} 
	ret_int = 0;
	return ret_int;
}